plugins/rust-openid-token-plugin/src/main.rs

```diff
@@ -0,0 +1,261 @@
+use async_mutex::Mutex;
+use clap::error::{Error, ErrorKind};
+use clap::Parser;
+use reqwest::header::{HeaderName, HeaderValue};
+use reqwest::{header::CONTENT_TYPE, Client, StatusCode};
+use std::str::FromStr;
+use std::sync::Arc;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use tonic::{transport::Server, Response, Status};
+use url::Url;
+#[macro_use]
+extern crate serde_derive;
+
+/// Requests an access token from an OpenID server and then injects the access
+/// token as an header into the API being tested by mapi.
+/// See https://mayhem4api.forallsecure.com/docs/rewrite.html for more details
+/// about rewrite plugins
+#[derive(Parser, Clone)]
+pub struct Args {
+    /// The GRPC port that will be listened on for rewrite requests from the `mapi` program.
+    /// Specify 0 to have an open port auto-selected.
+    /// May be specified using environment variable MAPI_PLUGIN_PORT.
+    #[arg(long, default_value_t = 50051, env)]
+    mapi_plugin_port: u16,
+
+    /// The URL of the openid-connect `/token` endpoint from which access tokens
+    /// will be fetched
+    #[arg(long, env)]
+    oauth_token_url: Url,
+
+    /// This option provides the body of the POST request sent to the `/token`
+    /// endpoint in order to fetch an access token. The data will be urlencoded.
+    /// Multiple calls will be joined with the `&` symbol and the full body of
+    /// the HTTP request will be sent with a `content-type` of
+    /// `application/x-www-form-urlencoded` This switch is a minimalist copy of
+    /// curl's `switch of the same name.
+    #[arg(long = "data-urlencode", env, value_parser(try_parse_data_urlencode))]
+    oath_data_urlencodes: Vec<String>,
+
+    /// A header to be sent to on the  oauth enpoint
+    #[arg(
+        long = "header",
+        env = "MAPI_OAUTH_REQUEST_HEADER",
+        value_parser(try_parse_header)
+    )]
+    oauth_request_headers: Vec<(HeaderName, HeaderValue)>,
+
+    /// The name of the HTTP header which will have the access token attached
+    /// when reqeusts are sent to the API under test
+    #[arg(
+        long = "api_under_test_header_name",
+        default_value = "authorization",
+        env,
+        value_parser(try_parse_header_name)
+    )]
+    api_under_test_header_name: HeaderName,
+
+    /// A prefix to prepended to the access token and set as the value of the
+    /// header field from `api_under_test_header_name`
+    #[arg(long = "api_under_test_header_prefix", env)]
+    api_under_test_header_prefix: Option<HeaderValue>,
```

Comment:
It is probably save to default this to `"Bearer "` -- what do you think?

Reply:
I had it like that, but then I have actually seen headers like `authorization: <jwt>` in the wild and not infrequently.  Then users have to do `--api_under_test_header_prefix=""`, which is rather unwieldy.  I agree that the default would serve more people, but those needing no prefix would find it a little odd.

Reply:
Ok, that makes sense. I think I would include the `"Bearer "` in the example in the readme in that case (and maybe in the docs here too). Then users can at least see where it is commonly used, without needing to explicitly clear out the default value.

Reply:
WIll do.

Reply:
[Done](https://github.com/ForAllSecure/mapi-examples/pull/4/files#diff-83ccb9c9f30f957a078ff94afd053fa8cbccdc063bf5bfbfb4c40f774b37bc3bR90)